department-of-veterans-affairs/caseflow
# Use VACOLS name for VLJ assignment and validation

Resolves https://vajira.max.gov/browse/CASEFLOW-2155

### Description
Updated HearingDaySerializer to serialize first name and last name with the data from CachedUser table. Collect data in batch to avoid any performance hit related to joining of tables for each row. Updated ValidateJudgeSpreadsheet so it validates first and last name with VACOLS data.

### Acceptance Criteria
- [ ] Assigning VLJs to dockets with Build Schedule accepts the first and last name from VACOLS in the uploaded file (rather than the full name from Caseflow's users table).
- [ ] Validates that the CSS ID and user name from VACOLS match each other.
- [ ] The Hearings Schedule page and the CSV downloaded from the page populates the "VLJ" column with the name from VACOLS ("<snamel>, <snamef>")

### Testing Plan

1. Go to the console `make c` and set the current user
```ruby
RequestStore[:current_user] = User.system_user
```
2. Create a `User` using `FactoryBot` with the role of judge and a VACOLS record:
```ruby
n = "Leonidas Olson"
id = "BVASLOLSON"
j = FactoryBot.create(:user, :with_vacols_judge_record, full_name: n, css_id: id)
```
3. Associate the newly created user to a `HearingDay` listed in http://localhost:3000/hearings/schedule
```ruby
hd = HearingDay.find(72)
hd.update!(judge: j)
```
4. Find and change the name on the `VACOLS::Staff` record associated with the user: 
```ruby
VACOLS::Staff.find_by(sdomainid: j.css_id).update(snamel: "Gutierrez")
```
5. Cache and sync all the VACOLS users
```ruby
CachedUser.sync_from_vacols
```
6. Shadow user BVASYELLOW http://localhost:3000/test/users
7. Navigate to the hearings schedule http://localhost:3000/hearings/schedule
8. Find the row where the `HearingDay` you associated with the new judge is. The name should match the `VACOLS::Staff` and `CachedUser` records and not the `User` record.
9. Download the table as a CSV by clicking on "Download current view".
10. Open the CSV in excel; edit the sheet so you have 2 or 3 judges. Make sure you have the new `User` you created above (the name here should also match the `VACOLS::Staff` and `CachedUser` records and not the `User` record).
11. Change the name on the sheet to match the name on the `User` table and save the file as an excel worksheet.
12. Navigate to http://localhost:3000/hearings/schedule/build
13. Click on "Upload Files" 
14. Select Judge assignment.
15. Upload the excel sheet you saved above.
16. Click Continue.
17. An error like this should display:
```
These judges are not in the database: [["BVASLOLSON", "Olson, Leonidas"]]
```
18. Edit the name in the spreadsheet to match the name in the `VACOLS::Staff` and `CachedUser` records and save it.
19. Upload the new excel sheet and click continue; you should see the "We have assigned your judges" message and a preview of the changes Caseflow will make on confirmation.
